storage/cloudspanner/map_storage.go

```diff
@@ -0,0 +1,373 @@
+		tx, err := ms.begin(ctx, tree, false /* readonly */, stx)
+		if err != nil {
+			glog.Errorf("failed to mapStorage.begin(treeID=%d): %v", tree.TreeId, err)
+			return err
+		}
+		if err := f(ctx, tx); err != nil {
+			return err
+		}
+		if err := tx.flushSubtrees(); err != nil {
+			glog.Errorf("failed to tx.flushSubtrees(): %v", err)
+			return err
+		}
+		return nil
+	})
+	return err
+}
+
+// mapTX is a concrete implementation of the Trillian storage.MapStorage
+// interface.
+type mapTX struct {
+	// treeTX embeds the merkle-tree level transactional actions.
+	*treeTX
+
+	// ms is the MapStorage which begat this mapTX.
+	ms *mapStorage
+}
+
+// sthToSMR converts a spannerpb.TreeHead to a trillian.SignedMapRoot.
+func sthToSMR(sth *spannerpb.TreeHead) (trillian.SignedMapRoot, error) {
+	mapRoot, err := (&types.MapRootV1{
+		RootHash:       sth.RootHash,
+		TimestampNanos: uint64(sth.TsNanos),
+		Revision:       uint64(sth.TreeRevision),
+		Metadata:       sth.Metadata,
+	}).MarshalBinary()
+	if err != nil {
+		return trillian.SignedMapRoot{}, err
+	}
+
+	return trillian.SignedMapRoot{
+		MapRoot:   mapRoot,
+		Signature: sth.Signature,
+	}, nil
+}
+
+// LatestSignedMapRoot returns the freshest SignedMapRoot for this map at the
+// time the transaction was started.
+func (tx *mapTX) LatestSignedMapRoot(ctx context.Context) (trillian.SignedMapRoot, error) {
+	currentSTH, err := tx.currentSTH(ctx)
+	if err != nil {
+		glog.Errorf("failed to determine current STH: %v", err)
+		return trillian.SignedMapRoot{}, err
+	}
+	writeRev, err := tx.writeRev(ctx)
+	if err != nil {
+		glog.Errorf("failed to determine write revision: %v", err)
+		return trillian.SignedMapRoot{}, err
+	}
+	if got, want := currentSTH.TreeRevision+1, writeRev; got != want {
+		return trillian.SignedMapRoot{}, fmt.Errorf("inconsistency: currentSTH.TreeRevision+1 (%d) != writeRev (%d)", got, want)
+	}
+
+	// We already read the latest root as part of starting the transaction (in
+	// order to calculate the writeRevision), so we just return that data here:
+	return sthToSMR(currentSTH)
+}
+
+// StoreSignedMapRoot stores the provided root.
+// This method will return an error if the caller attempts to store more than
+// one root per map for a given map revision.
+func (tx *mapTX) StoreSignedMapRoot(ctx context.Context, root trillian.SignedMapRoot) error {
+	stx, ok := tx.stx.(*spanner.ReadWriteTransaction)
+	if !ok {
+		return ErrWrongTXType
+	}
+
+	writeRev, err := tx.writeRev(ctx)
+	if err != nil {
+		glog.Errorf("failed to determine write revision: %v", err)
+		return err
+	}
+
+	var r types.MapRootV1
+	if err := r.UnmarshalBinary(root.MapRoot); err != nil {
+		return err
+	}
+	sth := spannerpb.TreeHead{
+		TsNanos:      int64(r.TimestampNanos),
+		RootHash:     r.RootHash,
+		Metadata:     r.Metadata,
+		TreeId:       tx.treeID,
+		TreeRevision: writeRev,
+		Signature:    root.Signature,
+	}
+
+	m := spanner.Insert(
```

Comment:
This is personal preference, but I'm enjoying using [`InsertStruct`](https://godoc.org/cloud.google.com/go/spanner#InsertStruct) these days as it's much clearer.

Reply:
I wrote this to match the existing LogStorage style, but sounds like we could/should go back over and migrate all the files in here to use `InsertStruct` in a follow up.

Reply:
Add a TODO?

Reply:
TODONE.